apm-application-toolkit/apm-toolkit-trace/src/main/java/org/apache/skywalking/apm/toolkit/trace/CorrelationContext.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.skywalking.apm.toolkit.trace;
+
+import java.util.Optional;
+
+/**
+ * CorrelationContext is the interactive API for end user to put/set custom data.
+ */
+public class CorrelationContext {
+
+    /**
+     * Try to get the custom value from trace context.
+     *
+     * @return custom data value.
+     */
+    public static Optional<String> get(String key) {
+        return Optional.empty();
+    }
+
+    /**
+     * Setting the custom key/value into trace context.
+     *
+     * @return previous value if it exists.
+     */
+    public static Optional<String> set(String key, String value) {
+        return Optional.empty();
+    }
```

Comment:
Personally, I don't recommend to return `Optional<String>` in these two APIs, for `String`s, we(developers, our users are developers too) don't usually **JUST** care about the nullability, we also want to check whether it's empty/blank, with `Optional<String>`, we have to unwrap the `Optional`, and check `isEmpty`/`isBlank`:

```java
        Optional<String> s = CorrelationContext.get(CORRELATION_CONTEXT_KEY);
        if (s.isPresent() && Strings.isNotBlank(s.get())) {
            // ...
        }
```

OR

```java
        if (!Strings.isNullOrEmpty(CorrelationContext.get(CORRELATION_CONTEXT_KEY).orElse(""))) {
            
        }
```

My point is that in this specific situation, `Optional<String>` may not bring much convenience as it should have done, simply returning nullable `String` may do, in the contrast

Reply:
You're saying to unwrap the `Optional<String>`, so we don't need to use `get`(or `orElse`)  and `Strings.isNullOrEmpty` to double-check. It's looking good to me. @wu-sheng any suggestion?

Reply:
> You're saying to unwrap the `Optional<String>`, so we don't need to use `get`(or `orElse`) and `Strings.isNullOrEmpty` to double-check

Exactly

Reply:
As `Optional`, the point is avoiding NPE explicitly for API user. We could guarantee there is no EMPTY_STRING as the value, as we discussed above. Then user doesn't have `doublt-check` case. We also don't ask user to do `.orElse("")` and then `Strings.isNullOrEmpty`. Your demo codes are written by yourself in this way.

I am a little confusing about your conclusion.

Reply:
> We could guarantee there is no EMPTY_STRING as the value, as we discussed above. 

Since [we have reached a consensus](https://github.com/apache/skywalking/pull/4555#discussion_r396902387) that `null`/`""` values should trigger a removal, this concern is gone accordingly